src/liquid/exceptions.cr

```diff
@@ -1,19 +1,43 @@
-# exceptions.cr
-
 module Liquid
+  # Base class for all exceptions raised by Liquid.cr shard.
   class LiquidException < Exception
   end
 
+  # Exception raised on syntax errors.
   class SyntaxError < LiquidException
+    # Line number where the syntax error was found.
     property line_number : Int32 = -1
   end
 
+  class InvalidStatement < LiquidException
+  end
+
+  # Exception used for any non-fatal errors that can happen while rendering a liquid template.
   class InvalidExpression < LiquidException
+    def initialize(message : String)
+      super("Liquid error: #{message}")
+    end
   end
 
-  class InvalidStatement < LiquidException
+  # Error generated when a variable used in a template doesn't exists in the current context.
+  #
+  # This exception is never raised whatever the context error mode, to access it check `Context#errors`.
+  class UndefinedVariable < InvalidExpression
+    def initialize(var_name : String)
+      super("Variable \"#{var_name}\" not found.")
```

Comment:
Maybe `"Undefined variable: #{var_name}"` to match the undefined filter message :thinking: